packages/core/src/scripts/migrate-oz-cli-project.ts

```diff
@@ -0,0 +1,426 @@
+}
+
+async function getProjectFile(): Promise<LegacyProjectFile> {
+  return JSON.parse(await fs.readFile(PROJECT_FILE, 'utf8'));
+}
+
+function isManifestFile(fileName: string): boolean {
+  const network = getNetworkName(fileName);
+  return isPublicNetwork(network) || isDevelopmentNetwork(network) || isUnknownNetwork(network);
+}
+
+function getNetworkName(fileName: string): string {
+  return path.basename(fileName, '.json');
+}
+
+function isDevelopmentNetwork(network: string): boolean {
+  // 31337      => buidler evm
+  // 1337       => ganache
+  // 13+ digits => ganache also uses timestamps
+  return /^dev-(3?1337|\d{13,})$/.test(network);
+}
+
+function isUnknownNetwork(network: string): boolean {
+  return !isDevelopmentNetwork(network) && /^dev-\d+$/.test(network);
+}
+
+function isPublicNetwork(network: string): boolean {
+  return ['mainnet', 'rinkeby', 'ropsten', 'kovan', 'goerli'].includes(network);
+}
+
+function getNewManifestLocation(oldName: string): string {
+  return isUnknownNetwork(oldName) ? oldName.replace('dev', 'unknown') : oldName;
+}
+
+async function writeJSONFile(location: string, data: unknown): Promise<void> {
+  await fs.writeFile(location, JSON.stringify(data, null, 2));
+}
+
+function getExportData(oldManifest: LegacyManifest): ExportData {
+  // we flexibilize the typing to allow required fields deletion
+  const manifestsMigrationData: Partial<LegacyManifest> = oldManifest;
+  delete manifestsMigrationData.proxyAdmin;
+  delete manifestsMigrationData.contracts;
+  delete manifestsMigrationData.solidityLibs;
+  return manifestsMigrationData as ExportData;
+}
+
+function updateManifest(oldManifest: LegacyManifest): ManifestData {
+  const proxyAdmin = oldManifest.proxyAdmin.address;
+
+  if (proxyAdmin === undefined) {
+    throw new Error('Legacy manifest does not have admin address');
+  }
+
+  if (Object.keys(oldManifest.solidityLibs).length > 0) {
+    throw new Error('Legacy manifest links to external libraries which are not yet supported');
+  }
+
+  return {
+    manifestVersion: '3.1',
+    impls: transformImplementations(oldManifest.contracts),
+    admin: {
+      address: proxyAdmin,
+    },
+  };
+}
+
+function transformImplementations(contracts: LegacyContracts): Record<string, ImplDeployment> {
+  const impls: Record<string, ImplDeployment> = {};
+  for (const contractName in contracts) {
+    const contract = contracts[contractName];
+    if (contract.deployedBytecodeHash === undefined) {
+      continue;
+    } else {
+      impls[contract.deployedBytecodeHash] = transformImplementationItem(contract);
+    }
+  }
+  return impls;
+}
+
+function transformImplementationItem(contract: LegacyContract): ImplDeployment {
+  if (contract.address === undefined) {
+    throw new Error('Could not find implementation address');
+  }
+
+  return {
+    address: contract.address,
+    layout: transformLayout(contract),
+  };
+}
+
+function transformLayout(contract: LegacyContract): StorageLayout {
```

Comment:
If either `storage` or `types` are undefined, I think it should be an error, rather than silently ignored.

Reply:
Why? can't a contract exist without any variables defined in it? They're also optional members of the types.

Reply:
In that case it should already be an empty array in the CLI layout, isn't it?

Reply:
Not sure, but I think we should be compliant with the types.

Reply:
Being compliant with the types means not ignoring the different possible types. Throwing an error is not ignoring.

The SDK types this as an optional property but it never emits it as undefined. See [`getStorageLayout`](https://github.com/OpenZeppelin/openzeppelin-sdk/blob/6c040b36052519f09569d6c40493707f2149b374/packages/lib/src/validations/Storage.ts#L10) and [`StorageLayoutInfo`](https://github.com/OpenZeppelin/openzeppelin-sdk/blob/6c040b36052519f09569d6c40493707f2149b374/packages/lib/src/validations/Storage.ts#L46-L49). If we find `undefined`, it's an error and we should abort the migration.